src/unique_combinations.rs

```diff
@@ -0,0 +1,112 @@
+use std::fmt;
+
+/// An iterator to iterate through all the `n`-length combinations in an iterator.
+///
+/// See [`.unique_combinations()`](../trait.Itertools.html#method.unique_combinations) for moref information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct UniqueCombinations<I: Iterator> {
+    len: usize,
+    position: Vec<usize>,
+    pool: Vec<I::Item>,
+    first: bool,
+    next_none: bool,
+}
+
+impl<I> fmt::Debug for UniqueCombinations<I>
+where
+    I: Iterator + fmt::Debug,
+    I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(Combinations, len, position, pool, first);
+}
+
+/// Create a new `UniqueCombinations` from a iterator with clonable and sorable Items.
+pub fn unique_combinations<I>(iter: I, len: usize) -> UniqueCombinations<I>
+where
+    I: Iterator,
+    I::Item: Ord,
+{
+    let mut pool: Vec<_> = iter.collect();
+    pool.sort_unstable();
+    UniqueCombinations {
+        len,
+        position: (0..len).collect(),
+        pool,
+        first: true,
+        next_none: false, // only used on iterators with 0 length
+    }
+}
+
+impl<I> UniqueCombinations<I>
+where
+    I: Iterator,
+    I::Item: Ord + Clone,
+{
+    #[inline]
+    fn generate(&self) -> Option<Vec<I::Item>> {
+        Some(
+            self.position
+                .iter()
+                .map(|n| self.pool[*n].clone())
+                .collect(),
+        )
+    }
+}
+
+impl<I> Iterator for UniqueCombinations<I>
+where
+    I: Iterator,
+    I::Item: Clone + Ord,
+{
+    type Item = Vec<I::Item>;
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
I may be missing something, but shouldn't `next` do something very similar to `Combinations::next`? (I.e. incrementing the rightmost index, possibly skipping values if the referenced values would lead to duplicates.)

If so, could we implement it similarly to `Combinations::next`?

Reply:
The difference is that combinations only bumps based on the value of the indices, and not the values of the pool.

Reply:
I agree that the first `if` and the `else if` should look the same. However the example there with the 2 `loop`-s that are just used as `jump` statements. I find that code a lot harder to read. 

My version shows the difference that either you bump the back index, or you find what index that can be bumped and sets the next ones after that index to the values following that index. 
If you trace your `if`-s you will see that they mostly run in one of these cases rarely bought (from what i found it was just to exit the `'outer` loop). Lines like line 35 is only useful in bumping the back number, as it will be overwritten in the `for` loop in the other cases.

Reply:
Hi there! Thanks a lot for bringing the `if`s in line - I really appreciate that and I think it looks better. Also very nice that you got rid of the `done` field.

I really try to grasp the algorithm and understand the distinct cases. Maybe you can help me with the following questions:

* Lines 77/78 (`lastpos` and `val`) seem to be analogous to lines 93/94 (`i` and `current`). If this is the case, would it be more readable if we called them the same, used the same logic wrt subtracting 1 from `i`/`lastpos`, and avoided the `mut`ability in the second case?
* Related: Is `for j in lastpos + 1..pool_len {...}` (line 79 and following) analogous to `while next == current` (line 96 and following)? If so, should we implement them in a similar fashion?

If all of the above are true:
* Do we then have the situation that the `else`-branch (starting at line 91) is basically a special case (avoiding setting the next one's indices) of the more general case (starting at 75)? If this is the case, would it make sense to leave out this special case (as `Combinations::next` does)?

Reply:
first point:
They look alike but the line 77 one has a i that is a count to count backwards. The other i is the position in the pool that contains the next item that is to be.I think we would get some benefit of calling them what they actually are, instead of something that is ambiguous.

Reply:
Hi again. I took your PR and did some syntactic adjustments (see [here](https://github.com/meltinglava/itertools/compare/ed0de3e383e09519cce7eeb6d141e48509771cbe...phimuemue:uniq_comb) for the individual steps).

In the end, I arrived at something where both cases actually use the same code, so we may be able to use similar code, and possibly could avoid special cases here.

Reply:
I found the solution now. Thanks for the branch.